wallet/walletrpc.c

```diff
@@ -160,6 +160,147 @@ static struct command_result *broadcast_and_wait(struct command *cmd,
+		enum address_parse_result res;
+
+		/* output format: {destination: amount} */
+		if (t->type != JSMN_OBJECT)
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+					    "The output format must be "
+					    "{destination: amount}");
+
+		res = json_to_address_scriptpubkey(cmd,
+						   chainparams,
+						   buffer, &t[1],
+						   &destination);
+		if (res == ADDRESS_PARSE_UNRECOGNIZED)
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+					    "Could not parse destination address");
+		else if (res == ADDRESS_PARSE_WRONG_NETWORK)
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+					    "Destination address is not on network %s",
+					    chainparams->network_name);
+
+		amount = tal(tmpctx, struct amount_sat);
+		if (!json_to_sat_or_all(buffer, &t[2], amount))
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+					    "'%.*s' is a invalid satoshi amount",
+					    t[2].end - t[2].start, buffer + t[2].start);
+
+		*outputs[i] = new_tx_output(*outputs, *amount,
+					    cast_const(u8 *, destination));
+		*out_len += tal_count(destination);
+
+		/* In fact, the maximum amount of bitcoin satoshi is 2.1e15.
+		 * It can't be equal to/bigger than 2^64.
+		 * On the hand, the maximum amount of litoshi is 8.4e15,
+		 * which also can't overflow. */
+		/* This means this destination need "all" satoshi we have. */
+		if (amount_sat_eq(*amount, AMOUNT_SAT(-1ULL))) {
+			if (outputstok->size > 1)
+				return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+						    "outputs[%zi]: this destination wants"
+						    " all satoshi. The count of outputs"
+						    " can't be more than 1. ", i);
+			(*utx)->wtx->all_funds = true;
+			/* `AMOUNT_SAT(-1ULL)` is the max permissible for `wallet_select_all`. */
+			(*utx)->wtx->amount = *amount;
+			break;
+		}
+
+		if (!amount_sat_add(&(*utx)->wtx->amount, (*utx)->wtx->amount, *amount))
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+					    "outputs: The sum of first %zi outputs"
+					    " overflow. ", i);
+	}
+	return NULL;
+}
+
+static struct command_result *create_tx(struct command *cmd,
+					struct unreleased_tx **utx,
+					struct bitcoin_tx_output **outputs,
+					const struct utxo **chosen_utxos,
+					size_t out_len,
+					u32 *feerate_per_kw,
+					u32 minconf,
+					u32 locktime)
+{
+	struct command_result *result;
+	u32 maxheight;
+	struct pubkey *changekey;
+
+	maxheight = minconf_to_maxheight(minconf, cmd->ld);
+
+	if (chosen_utxos)
+		result = wtx_from_utxos((*utx)->wtx, *feerate_per_kw,
+					out_len, maxheight,
+					chosen_utxos);
+	else
+		result = wtx_select_utxos((*utx)->wtx, *feerate_per_kw,
+					  out_len, maxheight);
+
+	if (result)
+		return result;
+
+	/* Because of the max limit of AMOUNT_SAT(-1ULL),
+	 * `(*utx)->wtx->all_funds` won't change in `wtx_select_utxos()` */
+	if ((*utx)->wtx->all_funds)
+		outputs[0]->amount = (*utx)->wtx->amount;
+
+	/* Add the change as the last output */
+	if (!amount_sat_eq((*utx)->wtx->change, AMOUNT_SAT(0))) {
```

Comment:
Does wally not set `wtx->change` if it is below dust? If that's not the case we should probably compare against the dust limit in `chainparams`.

Reply:
`wtx_select_utxos` does the dust comparison and sets this to zero if it's less than that, so this is an accurate check.